### PR TITLE
Add xbitinfo

### PR DESCRIPTION
**Pull request by rsignell-usgs**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by rsignell-usgs** [comment]
@observingClouds, I hope you are okay being a maintainer here!

**Comment by observingClouds** [comment]
Thank you so much @rsignell-usgs to put this together! I'm happy to be maintainer.

**Comment by ocefpaf** [comment]
At the moment this is failing with:

```
import: 'xbitinfo'
    import xbitinfo
  File "/opt/conda/conda-bld/xbitinfo_1661171801494/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.10/site-packages/xbitinfo/__init__.py", line 4, in <module>
    from .bitround import jl_bitround, xr_bitround
  File "/opt/conda/conda-bld/xbitinfo_1661171801494/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.10/site-packages/xbitinfo/bitround.py", line 4, in <module>
    from .xbitinfo import _jl_bitround, get_keepbits
  File "/opt/conda/conda-bld/xbitinfo_1661171801494/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.10/site-packages/xbitinfo/xbitinfo.py", line 12, in <module>
    jl = Julia(compiled_modules=False, debug=False)
  File "/opt/conda/conda-bld/xbitinfo_1661171801494/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.10/site-packages/julia/core.py", line 513, in __init__
    self._call("const PyCall = Base.require({0})".format(PYCALL_PKGID))
  File "/opt/conda/conda-bld/xbitinfo_1661171801494/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.10/site-packages/julia/core.py", line 549, in _call
    self.check_exception(src)
  File "/opt/conda/conda-bld/xbitinfo_1661171801494/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.10/site-packages/julia/core.py", line 603, in check_exception
    raise JuliaError(u'Exception \'{}\' occurred while calling julia code:\n{}'
julia.core.JuliaError: Exception 'ArgumentError' occurred while calling julia code:
const PyCall = Base.require(Base.PkgId(Base.UUID("438e738f-606a-5dbb-bf0a-cddfbfd45ab0"), "PyCall"))
Tests failed for xbitinfo-0.0.2-py_0.tar.bz2 - moving package to /opt/conda/conda-bld/broken
```

I don't know enough Julia to debug this. @observingClouds do you have any idea of what may be happening here? A Julia versions mismatch maybe?

**Comment by ocefpaf** [comment]
@ngam and @isuruf I'm quite lost when mixing Julia and Python. Do you have any advice here that I should follow?

**Comment by observingClouds** [comment]
@ocefpaf , it looks to me that we might not call the [PostInstallCommand](https://github.com/observingClouds/xbitinfo/blob/02c22c39d8a630286c97f28c265b319a3c3744c7/setup.py#L22-L27) which ensures that the Julia package dependencies are getting installed. `PyCall` is one of the Julia packages that needs to be installed with the [Julia package manager](https://github.com/observingClouds/xbitinfo/blob/main/install_julia_packages.jl). Maybe there is a way to install those with conda as well?

**Comment by ocefpaf** [comment]
> Maybe there is a way to install those with conda as well?

I don't think we have pycall. The number of julia packages in conda-forge is quite small. Julia has been moving target for a while and we left it on ice. We should probably revisit this now that Julia is more stable.

The PySR package also wraps a Julia module and this is how they do it: https://github.com/MilesCranmer/PySR/blob/master/pysr/julia_helpers.py

It is the most successful way of packaging Python-Julia wrappers at the moment. Maybe it is worth taking a look on what they do and see if we can port it to xbitinfo.

**Comment by ocefpaf** [comment]
> A potential solution for you is to add a step for the user to initiate the package with something like `module.install()` like the linked pysr.

I believe this is the best we can do. But I thought that the current setup should work regardless :-/

**Comment by ocefpaf** [comment]
> Until we hit that roadblock in pysr-feedstock in [conda-forge/pysr-feedstock#41](https://github.com/conda-forge/pysr-feedstock/pull/41).

Thanks! Glad to know we are not the only ones suffering from this.

@observingClouds are you willing to try the pysr approach? We can do that as a patch here first, see if that works, and then sending upstream.

**Comment by rsignell-usgs** [comment]
Or @aaronspring?

**Comment by observingClouds** [comment]
Thanks @rsignell-usgs and @ocefpaf for all the research and interest in making packaging of xbitinfo possible. The current pysr-approach reminds me of [a draft](https://github.com/observingClouds/xbitinfo/blob/dc20ce8a42f7c751d37ce24fe845851c179ebcb0/bitinformation_pipeline/bitinformation_pipeline.py#L9-L20) we had in the first days of developing xbitinfo. It was way less sophisticated and was called with every import of the module. We switched to the `PostInstallCommand` because this is called automatically when using setuptools to install the package.

I can try implementing pysr-approach (thanks @ngam for figuring this out). It will take me some time though. I'm happy for any further contributions.

**Comment by mkitti** [comment]
Umm, so why not do the obvious thing? Let's package the Julia packages within conda-forge.

**Comment by mkitti** [comment]
Here's a strategy we could use to bootstrap this. We're going to create a fake local depot containing all the packages we need. Then we're going to copy the parts we are interested into `xbitinfo_depot`.

```julia
using Pkg

# Create a temporary directory and activate it as the current environment
tmpdir = mktempdir(; cleanup=false)
cd(tmpdir)

# Clear the depots, and create a fresh fake depot
empty!(DEPOT_PATH)
fakedepot = joinpath(tmpdir, "fakedepot")
push!(DEPOT_PATH, fakedepot)

# Create a depot we are going to package
mkdir("xbitinfo_depot")
xbitinfo_env = joinpath("xbitinfo_depot", "environments", "xbitinfo")
mkpath(xbitinfo_env)
Pkg.activate(xbitinfo_env)

# Add the packages of interest
Pkg.add("PyCall")
Pkg.add("BitInformation")
Pkg.add("StatsBase")

# Capture the packages and artifacts directory
mv(joinpath(fakedepot, "packages"), joinpath("xbitinfo_depot", "packages"))
 # Including the artifacts might make the package platform specific, otherwise someone would still need to do "Pkg.instantiate"
mv(joinpath(fakedepot, "artifacts"), joinpath("xbitinfo_depot", "artifacts"))
```

Now "xbitinfo_depot" should contain just the "packages", "environments" and maybe "artifacts" directory. Package the "xbitinfo_depot" directory into conda tarball.

In pre-link.sh or any of the scripts, append a line to startup.jl to add "xbitinfo_depot" to the list of depots.
```
$ mkdir -p $CONDA_PREFIX/share/julia/config
$ echo "push!(DEPOT_PATH, \"/path/to/xbitinfo_depot\")" >> $CONDA_PREFIX/share/julia/config/startup.jl
```

Alternatively, we could further modify `$JULIA_DEPOT_PATH` in activate.sh:
```
export JULIA_DEPOT_PATH="$JULIA_DEPOT_PATH:/path/to/xbitinfo_depot"
```

What should happen now is that Julia will look into `xbitinfo_depot` for packages. The following should then work.

```julia
julia> using Pkg

julia> Pkg.activate("xbitinfo"; shared=true) # pkg"activate @xbitinfo"
  Activating project at `/tmp/jl_FA0ezT/xbitinfo_depot/environments/xbitinfo`

julia> using PyCall, StatsBase, BitInformation
[ Info: Precompiling PyCall [438e738f-606a-5dbb-bf0a-cddfbfd45ab0]
[ Info: Precompiling StatsBase [2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91]
[ Info: Precompiling BitInformation [de688a37-743e-4ac2-a6f0-bd62414d1aa7]

julia> 
```

@ngam where should `xbitinfo_depot` go so that it will be packaged?
@MilesCranmer, you may be interested in trying this as well

File: recipes/xbitinfo/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+requirements:
+  host:
+    - python >=3.8
+    - pip
+    - setuptools >=30.3.0
+    - setuptools-scm
+    - setuptools_scm_git_archive
+    - pyjulia >=0.5.7
+    - julia
```
Comment (MilesCranmer) [inline recipes/xbitinfo/meta.yaml:25]:
```suggestion
```

(Already comes with `pyjulia`)

File: recipes/xbitinfo/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  run:
+    - python >=3.8
+    - xarray
+    - julia
```
Comment (MilesCranmer) [inline recipes/xbitinfo/meta.yaml:29]:
```suggestion
```

File: recipes/xbitinfo/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  imports:
+    - xbitinfo
+  commands:
+    - pip check
+  requires:
+    - pip
+
+about:
+  home: https://github.com/observingClouds/xbitinfo
+  summary: Retrieve information content and compress accordingly.
```
Comment (MilesCranmer) [inline recipes/xbitinfo/meta.yaml:49]: Seems way too generic a summary

File: recipes/xbitinfo/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    # extras::prefect
+    - prefect >=1.0.0
+
+test:
+  imports:
+    - xbitinfo
+  commands:
+    - pip check
+  requires:
+    - pip
```
Comment (MilesCranmer) [inline recipes/xbitinfo/meta.yaml:45]: Probably want to add a test that you can import and run your package, since PyJulia+conda can be a bit finicky.

**Comment by MilesCranmer** [comment]
@observingClouds @ngam @mkitti if interested, maybe we could work together on a parent Python package that basically has the contents of this file: https://github.com/MilesCranmer/PySR/blob/master/pysr/julia_helpers.py, but for an arbitrary Julia dependency? Then both of our packages could depend on it, rather than PySR and your package hosting that helper code internally.

**Comment by mkitti** [comment]
@MilesCranmer, we need to address the issue of how conda-forge will distribute Julia packages directly preferably without having the user manually call `install()`. The code I posted above gets us slightly closer to this.

That said, it would be good to abstract out the reusable parts of your code.

**Comment by mkitti** [comment]
> What is really needed? For example, is stuff like this needed?\n\n

packages and artifacts

File: recipes/xbitinfo/meta.yaml (hunk 1 above)
Comment (ocefpaf) [reply recipes/xbitinfo/meta.yaml]: We need this so the conda-forge bot can apply the proper pinnings. In a way, Julia is a direct dependency and all direct dependencies must be specified.

File: recipes/xbitinfo/install_julia.patch [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+-        develop.run(self)
+-        os.system(julia_install_command)
+-
+-
+-class PostInstallCommand(install):
+-    """Post-installation for installation mode."""
+-
+-    def run(self):
+-        install.run(self)
+-        os.system(julia_install_command)
```
Comment (ocefpaf) [inline recipes/xbitinfo/install_julia.patch:27]: @observingClouds this is not ready yet but there are a few strategies I'd like to discuss with you and, if you agree, upstream them. First one is that I removed the installation at setup time.

File: recipes/xbitinfo/install_julia.patch [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+@@ -0,0 +1,128 @@
++"""Functions for initializing the Julia environment and installing deps."""
++import warnings
++from pathlib import Path
++import os
++
++from ._version import __version__
++
++
++def install(julia_project=None, quiet=False):  # pragma: no cover
```
Comment (ocefpaf) [inline recipes/xbitinfo/install_julia.patch:70]: @observingClouds the changes here are a copy from the pysr approach to install the juilia dependencies.

File: recipes/xbitinfo/install_julia.patch [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+--- xbitinfo-0.0.2.orig/xbitinfo/xbitinfo.py	2022-07-11 17:37:19.000000000 -0300
++++ xbitinfo-0.0.2/xbitinfo/xbitinfo.py	2022-08-29 14:16:24.923007159 -0300
+@@ -9,6 +9,11 @@
+ 
+ from . import __version__
+ 
++from .julia_helpers import install
++already_ran = False
++if not already_ran:
++    already_ran = install()
```
Comment (ocefpaf) [inline recipes/xbitinfo/install_julia.patch:200]: This is the trick to install/activate if it was not run before. PySr does a better job on this front but my Knowledge of Julia Pkg is pretty much zero. I have no idea how to improve this. At moment I hope it this is enough.

File: recipes/xbitinfo/install_julia.patch [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+ from ._version import __version__
+-from .bitround import jl_bitround, xr_bitround
+-from .graphics import plot_bitinformation, plot_distribution
+-from .save_compressed import get_compress_encoding_nc, get_compress_encoding_zarr
+-from .xbitinfo import get_bitinformation, get_keepbits, get_prefect_flow
++from .julia_helpers import install
++# from .bitround import jl_bitround, xr_bitround
++# from .graphics import plot_bitinformation, plot_distribution
++# from .save_compressed import get_compress_encoding_nc, get_compress_encoding_zarr
++# from .xbitinfo import get_bitinformation, get_keepbits, get_prefect_flow
```
Comment (ocefpaf) [inline recipes/xbitinfo/install_julia.patch:57]: @observingClouds this is probably the most controversial change but I needed to import the `install()` function without triggering the imports that depend on Julia. A nicer refatocr upstream is probaly better than this.

**Comment by ocefpaf** [comment]
@observingClouds and @rsignell-usgs Linux is passing, Windows won't pass b/c we don't have Julia for that platform yet, and I have no idea how to fix macOS. Still, this is kind of a win b/c this PR installs a working xbitinfo!

**Comment by rsignell-usgs** [comment]
I only care about Linux anyway -- the primary use on JupyterHubs with conda-store!

**Comment by ocefpaf** [comment]
OK. Let's bag what we have and watch the Julia+conda-forge space for future improvements. Thanks @ngam for all the pointers!

**Comment by ocefpaf** [comment]
> Feel free to tag me in the feedstock and keep us updated on what you come up with.

Will do but we'll probably follow your lead on the 'depot' attempt.

**Pull request merged**